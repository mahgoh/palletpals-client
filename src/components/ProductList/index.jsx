import PropTypes from 'prop-types'
import Grid from '@/components/Grid'
import ProductListItem from '@/components/ProductListItem'

export default function ProductList({ products }) {
  if (!products) {
    return null
  }

  return (
    <Grid>
      {products.map((product) => (
        <ProductListItem key={product.id} product={product} />
      ))}
    </Grid>
  )
}

ProductList.propTypes = {
  products: PropTypes.arrayOf(
    PropTypes.exact({
      id: PropTypes.number.isRequired,
      name: PropTypes.string.isRequired,
      price: PropTypes.number.isRequired,
      image: PropTypes.string,
    })
  ).isRequired,
}
