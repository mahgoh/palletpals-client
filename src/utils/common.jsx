export function classNames(...classes) {
  return classes.filter(Boolean).join(' ')
}

export function productImageURL(imageId) {
  return `${GLOBAL.API_URL}/product-images/${imageId}`
}

export function parseDateTime(dateTimeString, locale = 'en') {
  const datetime = new Date(dateTimeString)
  const options = {
    year: 'numeric',
    month: '2-digit',
    day: '2-digit',
    hour: 'numeric',
    minute: '2-digit',
  }

  return datetime.toLocaleString(locale, options)
}

export function formatPrice(price) {
  return `CHF ${price.toFixed(2)}`
}

export function formatOrderId(id) {
  return id.toString().padStart(6, '0')
}
