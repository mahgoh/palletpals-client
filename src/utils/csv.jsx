export async function processCSV(file) {
  try {
    const csv = await readCSV(file)
    const parsed = parseCSV(csv)

    return parsed
  } catch (e) {
    throw e
  }
}

function readCSV(file) {
  return new Promise((resolve, reject) => {
    if (file === null) reject('No file selected')

    // Check if the file is an image.
    if (file.type && !file.type === 'text/csv') {
      console.log('File is not csv.', file.type, file)
      reject('File is not csv.')
    }

    const reader = new FileReader()
    reader.addEventListener('load', (event) => {
      resolve(event.target.result)
    })
    reader.readAsText(file)
  })
}

/**
 *
 * Target:
 * {
 *   "km": [10,20,30],
 *   "pallets": [1,2,3],
 *   "price": [[5,6,7],[9,11,13],[15,17,19]]
 * }
 * @param {string} csv
 * @returns
 */

function parseCSV(csv) {
  const lines = csv.split('\n')

  // how entries are separated
  const delimiter = csv.includes(',') ? ',' : ';'

  const head = lines.shift()

  // palletArray is the first row except the first column
  // integers
  const palletArray = head
    .split(delimiter)
    .slice(1)
    .map((p) => parseInt(p))

  // kmArray is the first column except the first row
  // integers
  const kmArray = lines.map(
    (line) => line.split(delimiter).map((km) => parseInt(km))[0]
  )

  // priceMatrix is the rest of the table as two dimensional array
  // floats
  const priceMatrix = lines.map((line) =>
    line
      .split(delimiter)
      .slice(1)
      .map((p) => parseFloat(p))
  )

  return {
    kmArray,
    palletArray,
    priceMatrix,
  }
}
