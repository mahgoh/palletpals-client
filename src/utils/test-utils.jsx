import { cloneElement } from 'react'
import { BrowserRouter } from 'react-router-dom'
import { I18nextProvider } from 'react-i18next'
import { render } from '@testing-library/react'
import i18n from '../i18n'

import { AuthProvider } from '@/services/auth'
import { AppearanceProvider } from '@/services/appearance'
import { CartProvider } from '@/services/cart'
import { LanguageProvider } from '@/services/language'
import { NotificationProvider } from '@/services/notification'

/**
 * Render component with router wrapper and option for specifying the current route.
 *
 * @param {JSX} component
 * @param {Object} options
 * @returns
 */
const renderWithRouter = (component, { route = '/' } = {}) => {
  window.history.pushState({}, 'Test page', route)

  return {
    ...render(component, { wrapper: BrowserRouter }),
  }
}

/**
 * Render component with router wrapper and i18next provider.
 * @param {JSX} component
 * @param {Object} options
 * @returns
 */
const renderWithRouterWithi18n = (component, options) => {
  const comp = cloneElement(component, {
    changeLanguage: (lng) => {
      i18n.changeLanguage(lng)
      rerender(<I18nextProvider i18n={i18n}>{comp}</I18nextProvider>)
    },
  })
  const defaultRender = renderWithRouter(
    <I18nextProvider i18n={i18n}>{comp}</I18nextProvider>,
    options
  )
  const { rerender } = defaultRender
  return defaultRender
}

const renderComplete = (component, options) => {
  return {
    ...renderWithRouterWithi18n(
      <AuthProvider>
        <AppearanceProvider>
          <LanguageProvider>
            <CartProvider>
              <NotificationProvider>{component}</NotificationProvider>
            </CartProvider>
          </LanguageProvider>
        </AppearanceProvider>
      </AuthProvider>,
      options
    ),
  }
}

/**
 * Render component with specified renderer
 *
 * @param {JSX} component
 * @param {Object} param1
 * @returns
 */
const customRender = (component, { renderer = 'default', ...args } = {}) => {
  switch (renderer) {
    case 'router':
      return renderWithRouter(component, { ...args })
    case 'router-i18n':
      return renderWithRouterWithi18n(component, { ...args })
    case 'complete':
      return renderComplete(component, { ...args })
    case 'default':
    default:
      return {
        ...render(component),
      }
  }
}

export * from '@testing-library/react'
// export { default as userEvent } from '@testing-library/user-event'
export { customRender as render }
