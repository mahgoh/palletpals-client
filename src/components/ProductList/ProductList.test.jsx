import React from 'react'
import { render } from '@/utils/test-utils'
import ProductList from '.'

describe('product list', () => {
  it('should render', () => {
    const products = [
      {
        id: 1,
        name: 'DeWalt Saw Machine',
        price: 249,
      },
      {
        id: 2,
        name: 'Cover Suspension System',
        price: 89,
      },
    ]
    const { container } = render(<ProductList products={products} />, {
      renderer: 'router',
    })

    expect(container).not.toBeFalsy()
    expect(container).toMatchSnapshot()
  })

  it('should not render if products is falsy', () => {
    const { container } = render(<ProductList products={null} />, {
      renderer: 'router',
    })

    expect(container).toBeEmptyDOMElement()
  })

  it('should not render if products is undefined', () => {
    const { container } = render(<ProductList products={undefined} />, {
      renderer: 'router',
    })

    expect(container).toBeEmptyDOMElement()
  })
})
