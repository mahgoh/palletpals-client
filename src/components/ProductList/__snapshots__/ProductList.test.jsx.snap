// Vitest Snapshot v1

exports[`product list > should render 1`] = `
<div>
  <div
    class="grid grid-cols-1 gap-y-10 gap-x-6 sm:grid-cols-2 lg:grid-cols-3 xl:grid-cols-4 xl:gap-x-8"
  >
    <a
      class="group "
      href="/product/1"
    >
      <div
        class="aspect-w-1 aspect-h-1 w-full overflow-hidden bg-gray-100 dark:bg-gray-800 xl:aspect-w-7 xl:aspect-h-8"
      />
      <h3
        class="mt-4 text-sm text-gray-700 dark:text-gray-300"
      >
        DeWalt Saw Machine
      </h3>
      <p
        class="mt-1 text-lg font-medium text-gray-900 dark:text-gray-100"
      >
        CHF 249.00
      </p>
    </a>
    <a
      class="group "
      href="/product/2"
    >
      <div
        class="aspect-w-1 aspect-h-1 w-full overflow-hidden bg-gray-100 dark:bg-gray-800 xl:aspect-w-7 xl:aspect-h-8"
      />
      <h3
        class="mt-4 text-sm text-gray-700 dark:text-gray-300"
      >
        Cover Suspension System
      </h3>
      <p
        class="mt-1 text-lg font-medium text-gray-900 dark:text-gray-100"
      >
        CHF 89.00
      </p>
    </a>
  </div>
</div>
`;
